### ISSUE TITLE
json explosion false positive for large repos

### ISSUE THREAD

**Issue by damithc** [HubTurbo/HubTurbo#821]
Tried to download rails/rails. It triggered json explosion detection.
```
69091573 24 Jul 2015 10:57:42 INFO UI [pool-4-thread-56] - Triggered event UpdateRateLimitsEvent
69092626 24 Jul 2015 10:57:43 INFO GitHubRepo [pool-4-thread-44] - rails/rails | Unable to get collaborators: Must have push access to view repository collaborators. (403)
69092626 24 Jul 2015 10:57:43 INFO DownloadRepoTask [pool-4-thread-44] - rails/rails | Downloaded 20989 issue(s), 30 label(s), 42 milestone(s), 0 user(s)
69092626 24 Jul 2015 10:57:43 INFO UpdateService [pool-4-thread-58] - Updating rails/rails/collaborators with ETag null
69092626 24 Jul 2015 10:57:43 INFO UpdateService [pool-4-thread-60] - Updating rails/rails/milestones with ETag null
69092627 24 Jul 2015 10:57:43 INFO UpdateService [pool-4-thread-44] - Updating rails/rails/issues with ETag null
69092627 24 Jul 2015 10:57:43 INFO UpdateService [pool-4-thread-62] - Updating rails/rails/labels with ETag null
69092912 24 Jul 2015 10:57:44 WARN UpdateService [pool-4-thread-58] - UserUpdateService: error getting updated items (403)
69092912 24 Jul 2015 10:57:44 INFO UpdateUsersTask [pool-4-thread-58] - rails/rails | 0 user(s)) changed
69092930 24 Jul 2015 10:57:44 INFO UpdateService [pool-4-thread-62] - rails/rails/labels: 200
69093025 24 Jul 2015 10:57:44 INFO UpdateService [pool-4-thread-60] - rails/rails/milestones: 200
69093190 24 Jul 2015 10:57:44 INFO UpdateService [pool-4-thread-44] - rails/rails/issues: 200
69093257 24 Jul 2015 10:57:44 INFO UpdateService [pool-4-thread-62] - rails/rails/labels | page 0: 30 items
69093257 24 Jul 2015 10:57:44 INFO UpdateService [pool-4-thread-62] - New ETag for rails/rails/labels: Optional[2899ed6a07e2f7beb371c398c3d64398]
69093257 24 Jul 2015 10:57:44 INFO UpdateLabelsTask [pool-4-thread-62] - rails/rails | 30 label(s)) changed: [regression, actionmailer, actionpack, activemodel, activerecord, activesupport, railties, needs feedback, asset pipeline, third party issue, docs, i18n, actionview, Important, engines, onhold, JRuby, stale, routing, attached PR, rubinius, PostgreSQL, openacademy, security, GSoC, activejob, release blocker, enum, With reproduction steps, pinned]
69094926 24 Jul 2015 10:57:46 INFO UpdateService [pool-4-thread-60] - rails/rails/milestones | page 0: 42 items
69094926 24 Jul 2015 10:57:46 INFO UpdateService [pool-4-thread-60] - New ETag for rails/rails/milestones: Optional[dc14125d70c89017d74668a60d598e5a]
69094941 24 Jul 2015 10:57:46 INFO UpdateMilestonesTask [pool-4-thread-60] - rails/rails | 42 milestone(s)) changed: [3.1.4, 4.2.4, 4.1.13, 3.2.3, 3.1.5, 3.0.13, Rails 3.1, Rails 3.2, 4.1.9, 4.1.8, 4.1.6, 3.1.3, 4.0.0, 3.2.14, 5.0.0, 3.2.1, 3.2.6, 3.2.4, 4.0.1, 3.1.1, 3.2.9, 3.2.2, 3.2.15, 3.2.13, 3.2.7, 3.0.11, 4.0.4, 4.2.3, 3.2.10, 3.2.8, 4.1.12, Rails 3.0.8, 3.0.9, 3.0.10, 4.1.0, 4.1.2, 4.0.6, 4.0.10, 4.0.13, 4.2.0, 4.2.1, 4.1.10]
69095379 24 Jul 2015 10:57:46 INFO UpdateService [pool-4-thread-44] - rails/rails/issues | page 0: 42 items
69095380 24 Jul 2015 10:57:46 INFO UpdateService [pool-4-thread-44] - New ETag for rails/rails/issues: Optional[3a698f3abb77d5b0e624ebf91cbfd524]
69095380 24 Jul 2015 10:57:46 INFO UpdateIssuesTask [pool-4-thread-44] - rails/rails | 42 issue(s)) changed: [#21013 Brian davis, #21012 [close #18878] Proxy `rake` tasks through `rails`, #21011 wrap_parameters not including GET parameters, #21010 Make assert_enqueued_with and assert_performed_with returns the matched job, #21009 Issue with rake db:create/setup connecting to rails config/database.yml, #21008 Add documentation to get a running custom base controller, #21007 Encode the email address as prescribed in RFC 6068 section 2., #21006 Under concurrent load Dependencies.interlock might miss file changes, #21005 Docs: Update options for add_reference, #21004 to_partial_path namespaced model - wrong path, #21003 Allow generating plugins for API applications, #21002 ActiveRecord source plus inverse_of possibly broken in rails 4.2.x , #21001 json installation error , #21000 Update and fix forbidden attributes tests, #20999 Rename the enum_prefix and enum_suffix options to _prefix and _suffix, #20998 where.not doesn't work correctly with polymorphic columns, #20997 Changed id-writer to save join table records based on association pri…, #20996 Consistently protect primary key attribute methods against undefined primary keys, #20995 Save many-to-many objects based on association primary key, #20994 ActiveRecord transaction mixes objects up after rollback, #20993 Fixes documentation typo., #20992 Don't fail when checking dependencies in bin/setup script, #20990 AC::Parameters no longer being a Hash has broken find_or_initialize_by, #20989 Don't apply locking around basic #load / #require, #20988 ActiveRecord has_one associations not working after deserialization, #20961 Add dev caching toggle / server options, #20946 Freeze string literals when not mutated., #20933 in_batches using ids, #20928 We need stricter locking before we can unload, #20904 Add wildcard template dependencies., #20830 Refactored the enqueue methods for delayed supporting backends, #20800 Make ActiveJob locale aware, #20799 ActiveJob ignores locale during serialization, #20646 Mailer preview request model, #20609 Add documentation to get a running custom base controller, #20123 Fix for irregular inflection inconsistency, #20116 ActiveJob - log enqueued message only after the job was successfully enqueued, #20070 Adding max_retry to ActiveJob, #19793 Can't catch deserialization error when sending email with ActiveJob, #19456 Ignore index name in `index_exists?` when not passed a name to check for, #19256 Rails console doesn't use DATABASE_URL environment variable, #16983 ActiveRecord's `.where.not()` improperly negates polymorphic associations]
69095382 24 Jul 2015 10:57:46 INFO UpdateModelTask [pool-4-thread-56] - rails/rails | Updated model with 20991 issue(s), 30 label(s), 42 milestone(s), 0 user(s)
```

**Comment by damithc** [comment]
The backing up of the log doesn't seem to help much as it doesn't even contain proof of explosion. We might need to back up the json file as well.

**Comment by damithc** [comment]
It goes into an infinite loop or trying to download, trigger fired, re-attempt download :-p

**Comment by acruis** [comment]
The redownload loop is intended behaviour. I agree that we need to backup the json file tho. Will do it in a PR.

**Comment by damithc** [comment]
Re-downloading is good, except in cases like this where the trigger fires wrongly and repeatedly :-) If we cannot fix the explosion issue, and there is no sure fire way of detecting the explosion, then we need to limit the re-download to one attempt?

**Comment by acruis** [comment]
We'll just have to relax the acceptance criterion. Currently we're triggering a redownload if the file is more than 0.75KB per issue, which works okay with `TEAMMATES/repo`, but we didn't try with the bigger ones like Homebrew or rails. We can increase this limit.

The reason why we're not doing a simple file size ratio check before and after is that it triggers false positives on smaller repos, where a new issue can mean the repo doubling in size.

**Comment by damithc** [comment]
If not difficult to implement, a count limit is always a good thing. Say, if it fails two consecutive attempts, then chances of succeeding in the third attempt is low right? Repeating more times will only get in the way rather than helping.

**Comment by damithc** [comment]
Some other things to consider, to help troubleshoot this issue:
* Have line breaks in the json file. 
* Before writing an issue data to the file, do a sanity check on content and write the whole thing to the log file if the check fails.

**Comment by acruis** [comment]
What do you mean by sanity check?

**Comment by damithc** [comment]
like, how big it is, or even a regex check. This is assuming the explosion
happens because one of the fields contains a massive string of garbage
values.

**Comment by ndt93** [comment]
About the line break, we can set Gson to pretty print:

So instead of this:
```java
Gson gson = new Gson();
String json = gson.toJson(obj);
```

We can do this:

```java
Gson gson = new GsonBuilder().setPrettyPrinting().create();
String json = gson.toJson(obj);
```

Original tutorial is here: http://www.mkyong.com/java/how-to-enable-pretty-print-json-output-gson/

**Comment by acruis** [comment]
Yep. Probably not desirable for actual local store as it blows up the file size, but will be good for the copied version of the file (the one meant for debugging).

**Comment by acruis** [comment]
Closed in #824
